### ISSUE TITLE
lscpu command supported by ubunti distribution does not provide the complete set of information.

### ISSUE THREAD

**Issue by shyjupv** [hisilicon/hisilicon.github.io#2]
Problem details
-------------------
Ubuntu support the command "lscpu" which gives very valuable information about the target. 
The output obtained in X86 and ARM Arch are different, some of the key information is missing.
Please see the results copied below..
In Estuary (ARM)
----------------------
root@ltp-server:~# lscpu
Architecture:          aarch64
Byte Order:            Little Endian
CPU(s):                16
On-line CPU(s) list:   0-15
Thread(s) per core:    1
Core(s) per socket:    4
Socket(s):             4

In Ubuntu 14.04.2 (X86, Linux - 3.13.0-24-generic) 
-----------------------------------------------------------------
shyju@HTSATBLR-SERVER01:~/caliper-1.02/caliper-dev$ lscpu 
Architecture:          x86_64                             
CPU op-mode(s):        32-bit, 64-bit                     
Byte Order:            Little Endian                      
CPU(s):                4                                  
On-line CPU(s) list:   0-3                                
Thread(s) per core:    2                                  
Core(s) per socket:    2                                  
Socket(s):             1                                  
NUMA node(s):          1                                  
Vendor ID:             GenuineIntel                       
CPU family:            6                                  
Model:                 37                                 
Stepping:              5                                  
CPU MHz:               1199.000                           
BogoMIPS:              6384.15                            
Virtualization:        VT-x                               
L1d cache:             32K                                
L1i cache:             32K                                
L2 cache:              256K                               
L3 cache:              4096K                              
NUMA node0 CPU(s):     0-3                                
shyju@HTSATBLR-SERVER01:~/caliper-1.02/caliper-dev$       

Environment details:
--------------------------
Distribution : Estury 1.1
Board : D02
Linux version :  3.19.0-rc4+ (aarch64)

Reproduction of bug
-------------------------
 Run lscpu command

The missing information is key to get the target details. I think this command can be improved with more information. Please let me know.

**Comment by shamiali2008** [comment]
The lscpu is part of the util-linux package and uses kernel exported interfaces like /proc/cpuinfo, /sys/devices/system/cpuX etc to gather the cpu information and present it.
These exported interfaces differs on x86 and AARCH64 kernel. Hence the behaviour.

The below patch tries to add cache information 
https://patches.linaro.org/36620/
This patch is missing in the esturay 1.1 kernel, though this is included in 3.19 kernel.(I need to confirm this and has to see including this will print out the cache information by lscpu).

And patch to add CPU frequency looks like is not accepted:
http://linux-arm-kernel.infradead.narkive.com/Oq4y0pR9/patch-arm64-print-cpu-frequency-in-proc-cpuinfo

So whatever behaviour we are seeing is in-line with the kernel we are using, and if we are very particular about any information, can think of submitting a patch to include it in the mainline kernel.

**Comment by justin-zhao** [comment]
Hi, Shameer,
Could you try these patches, if they can work well, maybe you can submit them into estuary kernel tree, thank you.

**Comment by shamiali2008** [comment]
With the above mentioned cache information patch applied, lscpu on D02 now displays as below:

root@linaro-developer:~# lscpu
Architecture:          aarch64
Byte Order:            Little Endian
CPU(s):                16
On-line CPU(s) list:   0-15
Thread(s) per core:    1
Core(s) per socket:    4
Socket(s):             4
L1d cache:             32K
L1i cache:             48K
L2 cache:              1024K
root@linaro-developer:~#

**Comment by justin-zhao** [comment]
Great, Shameer, please submit this patch into https://github.com/hisilicon/linaro-kernel.git estuary branch.
And Shyju, is the above information from lscpu enough to you now?

**Comment by skdwriting** [comment]
Please ensure that the patches do not break upstream kernel. In that case,
let's add in a branch. Also, please check with latest hulk kernel and also
4.1. Because, we plan to rebase it.

User can take current base with temp patch to use for now.

What do you say shameer?

Thanks and Regards
-sanil.
"Make things simple and better!"

**Comment by justin-zhao** [comment]
Sanil, I really plan to upgrade estuary's kernel to 4.1, but maybe move to 4.1 directly, not rebase. because I'm afraid it's too difficult to rebase from 3.9-rc4 to 4.1, there are too many difference.

**Comment by skdwriting** [comment]
You want to simply move to hulk 4.1?

Thanks and Regards
-sanil.
"Make things simple and better!"

**Comment by justin-zhao** [comment]
Yes, I want to simply move to hulk 4.1 maybe before Dec, and add some patches into it to resolve what we solved in current version.

And Shyju, do you think we can close this issue now, if yes, please close it, thank you.
